src/NuGet.Core/NuGet.Packaging/Rules/NoRefOrLibFolderInPackageRule.cs

```diff
@@ -0,0 +1,53 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using NuGet.Common;
+
+namespace NuGet.Packaging.Rules
+{
+    class NoRefOrLibFolderInPackageRule : IPackageRule
+    {
+        public string MessageFormat { get; }
+
+        public NoRefOrLibFolderInPackageRule(string messageFormat)
+        {
+            MessageFormat = messageFormat;
+        }
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader package)
+        {
+            var files = package.GetFiles();
+            if (!files
+                    .Select(t => PathUtility.GetPathWithDirectorySeparator(t))
+                    .Any(t => t.StartsWith(PackagingConstants.Folders.Lib + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)
+                    || t.StartsWith(PackagingConstants.Folders.Ref + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)))
+            {
+                //if you can't find the ref and lib folder, then find the build folder
+                if (files
+                    .Select(t => PathUtility.GetPathWithDirectorySeparator(t))
+                    .Any(t => t.StartsWith(PackagingConstants.Folders.Build + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)))
+                {
+                    //if you can't find any folders other than native or any, raise an NU5127
+                    if (files
+                        .Select(t => PathUtility.GetPathWithDirectorySeparator(t))
```

Comment:
Since all usages of `files` use `.Select(t => PathUtility.GetPathWithDirectorySeparator(t)`, you should move it to when `files` is initialised, so the conversion only needs to happen once, not once per usage.